File: Haskell/src/Modelos/Mercado.hs
{-# LANGUAGE DeriveGeneric #-}

module Modelos.Mercado (Mercado(..), MercadoClass(..)) where

import Modelos.Jogador (Jogador(..), JogadorClass(..))

import GHC.Generics (Generic)
import Data.Aeson (ToJSON, FromJSON, encode, decode, eitherDecode)

class MercadoClass m where 
    getPrecoBM :: m -> Int
    getPrecoBG :: m -> Int
    getPrecoDV :: m -> Int

    comprarItem :: Jogador -> m -> String -> Jogador  

data Mercado = Mercado 
    { precoBombasMedias :: Int
    , precoBombasGrandes :: Int
    , precoDroneVisualizador :: Int
    } deriving (Show, Generic)

instance MercadoClass Mercado where
    getPrecoBM = precoBombasMedias
    getPrecoBG = precoBombasGrandes
    getPrecoDV = precoDroneVisualizador

    comprarItem jogador mercado item = case item of
        "1" -> 
            if getMoedas jogador >= getPrecoBM mercado
                then setMoedas (setBombasMedias jogador (getBombasMedias jogador + 1)) (getMoedas jogador - getPrecoBM mercado)
                else jogador
        "2" -> 
            if getMoedas jogador >= getPrecoBG mercado
                then setMoedas (setBombasGrandes jogador (getBombasGrandes jogador + 1)) (getMoedas jogador - getPrecoBG mercado)
                else jogador
        "3" -> 
            if getMoedas jogador >= getPrecoDV mercado
                then setMoedas (setDroneVisualizador jogador (getDroneVisualizador jogador + 1)) (getMoedas jogador - getPrecoDV mercado)
                else jogador
        naoExiste -> jogador  

instance ToJSON Mercado
instance FromJSON Mercado

File: Prolog/src/Modelos/Mercado.pl
:- module(mercado, [
    getPrecoBM/2, getPrecoBG/2, getPrecoDV/2,
    comprarItem/3
]).

:- use_module('./Jogador.pl').

:- dynamic mercado/3.

% mercado(PrecoBM, PrecoBG, PrecoDV)

% Getters

getPrecoBM(mercado(PrecoBM, _, _), PrecoBM).
getPrecoBG(mercado(_, PrecoBG, _), PrecoBG).
getPrecoDV(mercado(_, _, PrecoDV), PrecoDV).

% Compra de itens

comprarItem(Jogador, Mercado, "1", NovoJogador) :-
    getMoedas(Jogador, Moedas),
    getPrecoBM(Mercado, PrecoBM),
    Moedas >= PrecoBM,
    getBombasMedias(Jogador, BM),
    NovoBM is BM + 1,
    NovoMoedas is Moedas - PrecoBM,
    setBombasMedias(Jogador, NovoBM, TempJogador),
    setMoedas(TempJogador, NovoMoedas, NovoJogador).

comprarItem(Jogador, Mercado, "2", NovoJogador) :-
    getMoedas(Jogador, Moedas),
    getPrecoBG(Mercado, PrecoBG),
    Moedas >= PrecoBG,
    getBombasGrandes(Jogador, BG),
    NovoBG is BG + 1,
    NovoMoedas is Moedas - PrecoBG,
    setBombasGrandes(Jogador, NovoBG, TempJogador),
    setMoedas(TempJogador, NovoMoedas, NovoJogador).

comprarItem(Jogador, Mercado, "3", NovoJogador) :-
    getMoedas(Jogador, Moedas),
    getPrecoDV(Mercado, PrecoDV),
    Moedas >= PrecoDV,
    getDroneVisualizador(Jogador, DV),
    NovoDV is DV + 1,
    NovoMoedas is Moedas - PrecoDV,
    setDroneVisualizador(Jogador, NovoDV, TempJogador),
    setMoedas(TempJogador, NovoMoedas, NovoJogador).

comprarItem(Jogador, _, _, Jogador).